src/filesystem/nodes/metadata/mime_type.rs

```diff
@@ -0,0 +1,212 @@
+use crate::prelude::nodes::NodeName;
+
+
+#[derive(Default)]
+pub struct MimeGuesser {
+    name: Option<String>,
+    data: Vec<u8>,
+}
+
+impl MimeGuesser {
+    const MP3_RATES: [u32; 15] = [
+        0, 32000, 40000, 48000, 56000, 64000, 80000, 96000, 112000, 128000, 160000, 192000, 224000,
+        256000, 320000,
+    ];
+
+    const MP25_RATES: [u32; 15] = [
+        0, 8000, 16000, 24000, 32000, 40000, 48000, 56000, 64000, 80000, 96000, 112000, 128000,
+        144000, 160000,
+    ];
+
+    const SAMPLE_RATES: [u32; 3] = [44100, 48000, 32000];
+
+    pub fn with_name(mut self, name: NodeName) -> Self {
+        match name {
+            NodeName::Named(name) => self.name = Some(name.clone()),
+            NodeName::Root => {}
+        }
+        self
+    }
+
+    pub fn with_data(mut self, data: &[u8]) -> Self {
+        self.data.extend_from_slice(data);
+        self
+    }
+
+    pub fn guess_mime_type(&self) -> Option<mime::MediaType> {
+        self.pattern_match()
+            .or_else(|| self.algorithm_match())
+            .or_else(|| self.extension_match())
+    }
+
+    fn extension_match(&self) -> Option<mime::MediaType> {
+        let guess = mime_guess::get_mime_extensions_str(
+            self.name.as_ref().map_or("", |name| name.as_str()),
+        );
```

Comment:
I might be reading this wrong, but it looks like we are feeding this function a file name, but the docs say it expects a mime string (e.g. `application/zip` ) and then returns a list of extensions that that mime string is known to go by. The docs actually say it can panic if it is passed a string that isn't a mime string https://docs.rs/mime_guess/latest/mime_guess/fn.get_mime_extensions_str.html

Reply:
but doesn't `self.name` just come from the Nodename passed to `with_name()`? So it might be say `my_file.zip` that gets fed into that function, which is expecting `application/zip`?

Reply:
Ah yeah you are right - I messed up the functoin call. Should be `mime_guess::from_path`. Nice catch